docs/api-spec.md

```diff
@@ -0,0 +1,322 @@
+| `resourceVersion`    | string             | RECOMMENDED |
+| `generation`         | int64              | RECOMMENDED |
+| `namespace`          | string             | RECOMMENDED |
+| `generateName`**     | string             | RECOMMENDED |
+
+\* `creationTimestamp` MUST be populated by the implementation, in [RFC3339](https://tools.ietf.org/html/rfc3339).
+
+** If `generateName` is supported by the implementation, when it is specified at creation, it MUST be prepended to a random string and set as the `name`, and not set on the subsequent response.
+
+#### Spec
+
+| Field Name            | Field Type           | Requirement |
+|-----------------------|----------------------|-------------|
+| `params`              | `[]Param`            | REQUIRED    |
+| `taskSpec`            | `TaskSpec`           | REQUIRED    |
+| `workspaces`          | `[]WorkspaceBinding` | REQUIRED    |
+| `status`              | Enum:<br>- `""` (default)<br>- `"TaskRunCancelled"` | RECOMMENDED |
+| `timeout`             | string (duration)    | RECOMMENDED |
+| `serviceAccountName`^ | string               | RECOMMENDED |
+
+^ In the Kubernetes implementation, `serviceAccountName` refers to a Kubernetes `ServiceAccount` resource that is assumed to exist in the same namespace. Other implementations MAY interpret this string differently, and impose other requirements on specified values.
+
+#### Status
+
+| Field Name            | Field Type             | Requirement |
+|-----------------------|------------------------|-------------|
+| `conditions`          | see [#error-signaling] | REQUIRED    |
+| `startTime`           | string                 | REQUIRED    |
+| `completionTime`*     | string                 | REQUIRED    |
+| `steps`               | `[]StepState`          | REQUIRED    |
+| `observedGeneration`  | int64                  | RECOMMENDED |
+
+\* `startTime` and `completionTime` MUST be populated by the implementation, in [RFC3339](https://tools.ietf.org/html/rfc3339).
+
+## Status Signalling
+
+The Tekton Pipelines API uses the [Kubernetes Conditions convention](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties) to communicate status and errors to the user.
+
+`TaskRun`'s `status` field MUST have a `conditions` field, which must be a list of `Condition` objects of the following form:
+
+| Field                 | Type   |
+|-----------------------|--------|
+| `type`                | string |
+| `status`              | Enum:<br>- `"True"`<br>- `"False"`<br>- `"Unknown"` (default) |
+| `reason`              | string |
+| `message`             | string |
+| `severity`            | Enum:<br>- `""` (default)<br>- `"Warning"`<br>- `"Info"` |
+| `lastTransitionTime`* | string |
+
+\* If `lastTransitionTime` is populated by the implementation, it must be in [RFC3339](https://tools.ietf.org/html/rfc3339). Reporting `lastTransitionTime` is OPTIONAL.
```

Comment:
should we list REQUIRED, etc for the fields above as well? (e.g. i have no idea what severity is about, seems less important)